### PR TITLE
Contrastive Reranker/Reward model 

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

enables training of a contrastive reranker model based on GPT architecture and PEFT. Used for reranking the results of an embedding model.

**Collection**: [NLP]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/information_retrieval/megatron_gpt_reranker_finetuning.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections.abc import MutableMapping
+
+import torch.multiprocessing as mp
+from omegaconf.omegaconf import OmegaConf
+from pytorch_lightning.loggers import WandbLogger
+
+from nemo.collections.nlp.models.information_retrieval.megatron_gpt_embedding_model import MegatronGPTEmbeddingModel
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply examples/nlp/information_retrieval/megatron_gpt_reranker_finetuning.py:21]: can you resolve this

File: examples/nlp/information_retrieval/megatron_gpt_reranker_finetuning.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (cuichenx) [inline examples/nlp/information_retrieval/megatron_gpt_reranker_finetuning.py:1]: new file copyright year should be 2024

File: examples/nlp/information_retrieval/megatron_gpt_reranker_generate.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (cuichenx) [inline examples/nlp/information_retrieval/megatron_gpt_reranker_generate.py:1]: new file year should be 2024

File: nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (cuichenx) [inline nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py:1]: new file year should be 2024

File: nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+    HAVE_MEGATRON_CORE = True
+
+except (ImportError, ModuleNotFoundError):
+
+    HAVE_MEGATRON_CORE = False
+try:
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py:42]: can get rid of this try except block?

File: nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        local_eos_tensor = local_eos_tensor.contiguous()
+        global_eos_tensors = [
+            torch.zeros_like(local_eos_tensor) for _ in range(parallel_state.get_data_parallel_world_size())
+        ]
+        torch.distributed.all_gather(
+            global_eos_tensors, local_eos_tensor, group=parallel_state.get_data_parallel_group()
+        )
+        global_eos_tensors[parallel_state.get_data_parallel_rank()] = local_eos_tensor
+        global_eos_tensors = torch.cat(global_eos_tensors, dim=0)
+        return global_eos_tensors
```
Comment (cuichenx) [inline nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py:208]: these two functions don't use "self", i wonder if it's better to make them helper functions outside of this class?

File: nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+    def write_embeddings_to_file(self, outputs, output_file_path, d_idx):
+        hs = torch.cat(outputs['query_pos_doc_logit'], dim=0)
+        hs_npy = hs.float().numpy()
+        emb_fldr = f"{output_file_path}"
+        os.makedirs(emb_fldr, exist_ok=True)
+        with open(f"{output_file_path}/logits.ids", "w") as f:
+            for m in outputs['metadata']:
+                f.write(f"{m['query_id'].strip()} {m['doc_id']}\n")
+        np.save(f"{emb_fldr}/logits.npy", hs_npy)
+        return True
```
Comment (cuichenx) [inline nemo/collections/nlp/models/information_retrieval/megatron_gpt_reranker_model.py:317]: same as above

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 8]
```diff
@@ -1238,24 +1237,47 @@ def loss_func(output_tensor):
...
+                            {
+                                'avg': reduced_loss,
+                                'query_pos_doc_logit': loss_for_ub['query_pos_doc_logit'],
+                                'query_neg_doc_logit': loss_for_ub['query_neg_doc_logit'],
+                                'logit_diff': logit_diff,
+                            },
+                        )
+                    else:
+                        raise RuntimeError(f"Dict loss_for_ub has unknown key set {loss_for_ub.keys()}")
+
```
Comment (cuichenx) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1280]: do we still need this change? given that we implement loss_func in the reranker model?

File: nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py [hunk 9]
```diff
@@ -63,6 +70,137 @@ def mcore_register_adapters(self):
...
+                    if (
+                        torch.is_grad_enabled()
+                        and self.config.cpu_offloading
+                        and self.group_prefetch_offload_commit_async is not None
+                    ):
+                        hidden_states = self.group_prefetch_offload_commit_async(hidden_states)
+
+        # Final layer norm.
+        if self.post_process and self.post_layer_norm:
+            hidden_states = self.final_layernorm(hidden_states)
```
Comment (cuichenx) [inline nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py:195]: can this be replaced with super.forward()?

File: nemo/collections/nlp/modules/common/megatron/adapters/parallel_adapters.py [hunk 10]
```diff
@@ -373,6 +376,57 @@ class ParallelLinearAdapterConfig(AdapterConfig):
...
+        out_features: int,
+        input_is_parallel: bool = False,
+        model_parallel_config: Optional[ModelParallelConfig] = None,
+        **kwargs,
+    ):
+        super().__init__()
+        if model_parallel_config is None:
+            model_parallel_config = ModelParallelConfig()
+        self._sequence_parallel = model_parallel_config.sequence_parallel
+        model_parallel_config.sequence_parallel = False  # SP is irrelevant for the lora linear layer
```
Comment (cuichenx) [inline nemo/collections/nlp/modules/common/megatron/adapters/parallel_adapters.py:392]: can you update this comment

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 8 above)
Comment (arendu) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1280]: yes, we do the subclass `loss_func` is called in line 1238. This part resolves the loss form different data parallel ranks.

File: nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py (hunk 9 above)
Comment (arendu) [reply nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py:195]: ah, yes, will do!

**Review by cuichenx — APPROVED** [review]
LGTM

File: examples/nlp/information_retrieval/conf/megatron_gpt_embedder_tuning_config.yaml [hunk 11]
```diff
@@ -79,6 +79,7 @@ model:
   num_soft_negatives: 0 # Number of soft negatives to use for contrastive loss,it should be max(batch_size - 1), 0 means use hard negatives only
   use_all_possible_negatives: False # If True, use all possible negatives for contrastive loss, otherwise use num_soft_negatives, if num_soft_negatives is 0, use hard negatives only
   post_process: False # should be False.
+  apply_rope_fusion: False
```
Comment (cuichenx) [inline examples/nlp/information_retrieval/conf/megatron_gpt_embedder_tuning_config.yaml:82]: this change leads to the duplicate key

**Review by cuichenx — DISMISSED** [review]

**Review by cuichenx — APPROVED** [review]
LGTM. Looks like pipeline parallel to be verified and supported in a subsequent PR.
